Firestore/core/src/firebase/firestore/local/leveldb_mutation_queue.h

```diff
@@ -0,0 +1,152 @@
+/*
+ * Copyright 2019 Google
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef FIRESTORE_CORE_SRC_FIREBASE_FIRESTORE_LOCAL_LEVELDB_MUTATION_QUEUE_H_
+#define FIRESTORE_CORE_SRC_FIREBASE_FIRESTORE_LOCAL_LEVELDB_MUTATION_QUEUE_H_
+
+#if !defined(__OBJC__)
+#error "For now, this file must only be included by ObjC source files."
+#endif  // !defined(__OBJC__)
+
+#import <Foundation/Foundation.h>
+
+#include <set>
+#include <string>
+#include <vector>
+
+#import "Firestore/Source/Public/FIRTimestamp.h"
+
+#include "Firestore/core/src/firebase/firestore/auth/user.h"
+#include "Firestore/core/src/firebase/firestore/local/leveldb_key.h"
+#include "Firestore/core/src/firebase/firestore/model/document_key.h"
+#include "Firestore/core/src/firebase/firestore/model/document_key_set.h"
+#include "Firestore/core/src/firebase/firestore/model/types.h"
+#include "absl/strings/string_view.h"
+#include "leveldb/db.h"
+
+@class FSTLevelDB;
+@class FSTLocalSerializer;
+@class FSTMutation;
+@class FSTMutationBatch;
+@class FSTPBMutationQueue;
+@class FSTQuery;
+
+NS_ASSUME_NONNULL_BEGIN
+
+namespace firebase {
+namespace firestore {
+namespace local {
+
+/**
+ * Returns one larger than the largest batch ID that has been stored. If there
+ * are no mutations returns 0. Note that batch IDs are global.
+ */
+model::BatchId LoadNextBatchIdFromDb(leveldb::DB* db);
+
+class LevelDbMutationQueue {
+ public:
+  LevelDbMutationQueue(const auth::User& user,
+                       FSTLevelDB* db,
+                       FSTLocalSerializer* serializer);
+
+  void Start();
+
+  bool IsEmpty();
```

Comment:
In one of the previous ports, we decided that things accessing leveldb internally should not be marked `const`, even if they aren't mutating data stored in leveldb. Unfortunately, this bleeds out to the interface a bit, meaning that even the memory implementation of `IsEmpty()` can't be marked `const` since it will need to match the leveldb implementation.

Reply:
I think maybe the appropriate place will be the ported version of `FSTLevelDB`, since it's the common denominator for leveldb classes? Not sure though.

Reply:
Yeah that seems appropriate; but I suggested the interface, since, as you said, it "bleeds out to the interface a bit". Doesn't matter too much though.

(Scenario I'm "worried" about is me seeing the memory implementation and noticing the lack of const. So I add it there, and to the interface as well. But in this case, the compiler would notice and yell at me, forcing me to check out FSTLevelDB, so I'd eventually find the comment regardless.)